<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>ParsEval Summary</title>
    <link rel="stylesheet" type="text/css" href="parseval.css" />
    <script type="text/javascript" language="javascript" src="vendor/jquery.js"></script>
    <script type="text/javascript" language="javascript" src="vendor/jquery.dataTables.js"></script>
    <script type="text/javascript">
      $(document).ready(function() {
        $('#seqlist').dataTable( {
          "sScrollY": "400px",
          "bPaginate": false,
          "bScrollCollapse": true,
          "bSort": false,
          "bFilter": false,
          "bInfo": false
        });
      } );
    </script>
  </head>
  <body>
    <div id="content">
      <h1>ParsEval Summary</h1>
      <pre class="command">
Started:                03 Jan 2021, 02:54PM
Reference annotations:  /tmp/tmpok6hen4l/files/3/2/4/dataset_324b4f7b-e4b1-45d0-a543-be63ae960b43.dat
Prediction annotations: /tmp/tmpok6hen4l/files/b/4/0/dataset_b4066664-7777-4c30-9e37-820e44194617.dat
Executing command:      parseval --delta 10 --maxtrans 10 -w -f html -o /tmp/tmpok6hen4l/job_working_directory/000/46/working/dataset_1f07132b-8f99-4a03-8305-5c4c5a68089c_files /tmp/tmpok6hen4l/files/3/2/4/dataset_324b4f7b-e4b1-45d0-a543-be63ae960b43.dat /tmp/tmpok6hen4l/files/b/4/0/dataset_b4066664-7777-4c30-9e37-820e44194617.dat </pre>

      <h2>Sequences compared</h2>
      <p class="indent">Click on a sequence ID below to see comparison results for individual loci.</p>
      <table id="seqlist" class="indent">
        <thead>
          <tr>
            <th>Sequence</th>
            <th>Refr genes</th>
            <th>Pred genes</th>
            <th>Loci</th>
          </tr>
        </thead>
        <tbody>
        <tr><td><a href="Chr1/index.html">Chr1</a></td><td>4</td><td>4</td><td>4</td></tr>
        </tbody>
      </table>

      <h2>Gene loci <span class="tooltip">[?]<span class="tooltip_text">If a gene annotation overlaps with another gene annotation, those annotations are associated with the same gene locus. See <a target="_blank" href="http://aegean.readthedocs.io/en/latest/loci.html">this page</a> for a formal definition of a locus annotation.</span></span></h2>
      <table class="table_normal">
        <tr><td>shared</td><td>4</td></tr>
        <tr><td>unique to reference</td><td>0</td></tr>
        <tr><td>unique to prediction</td><td>0</td></tr>
        <tr><th class="right-align">Total</th><td>4</td></tr>
      </table>

      <h2>Reference annotations</h2>
      <table class="table_normal">
        <tr><td>genes</td><td>4</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per locus</td><td>1.000</td></tr>
        <tr><td>transcripts</td><td>5</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per locus</td><td>1.250</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per gene</td><td>1.250</td></tr>
      </table>

      <h2>Prediction annotations</h2>
      <table class="table_normal">
        <tr><td>genes</td><td>4</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per locus</td><td>1.000</td></tr>
        <tr><td>transcripts</td><td>5</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per locus</td><td>1.250</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per gene</td><td>1.250</td></tr>
      </table>

      <h2>Comparisons</h2>
      <table class="comparisons">
        <tr><th>Total comparisons</th><th>5</th></tr>
        <tr><td><a href="perfectmatches.html">(+)</a> perfect matches <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">Prediction transcripts (exons, coding sequences, and UTRs) line up perfectly with reference transcripts.</span></span></td><td>5 (100.0%)</tr>
        <tr class="cell_small"><td class="cell_indent">average length</td><td>3629.80 bp</td></tr>
        <tr class="cell_small"><td class="cell_indent">average # refr exons</td><td>7.60</td></tr>
        <tr class="cell_small"><td class="cell_indent">average # pred exons</td><td>7.60</td></tr>
        <tr class="cell_small"><td class="cell_indent">average refr CDS length</td><td>449.40 aa</td></tr>
        <tr class="cell_small"><td class="cell_indent">average pred CDS length</td><td>449.40 aa</td></tr>
        <tr><td>(+) perfect matches with mislabeled UTRs <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">5'/3' orientation of UTRs is reversed between reference and prediction, but a perfect match in all other aspects.</span></span></td><td>0 (0.0%)</tr>
        <tr><td>(+) CDS structure matches <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">Not a perfect match, but prediction coding sequence(s) line up perfectly with reference coding sequence(s).</span></span></td><td>0 (0.0%)</tr>
        <tr><td>(+) exon structure matches <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">Not a perfect match or CDS match, but prediction exon structure is identical to reference exon structure.</span></span></td><td>0 (0.0%)</tr>
        <tr><td>(+) UTR structure matches <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">Not a perfect match, CDS match, or exon structure match, but prediction UTRs line up perfectly with reference UTRs.</span></span></td><td>0 (0.0%)</tr>
        <tr><td>(+) non-matches <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">Differences in CDS, exon, and UTR structure.</span></span></td><td>0 (0.0%)</tr>
      </table>

      <h3>CDS structure comparison</h3>
      <table class="table_normal table_extra_indent">
        <tr><td>reference CDS segments</td><td>35</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>35 (100.0%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>0 (0.0%)</td></tr>
        <tr><td>prediction CDS segments</td><td>35</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>35 (100.0%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>0 (0.0%)</td></tr>
        <tr><td>Sensitivity</td><td>1.000</td></tr>
        <tr><td>Specificity</td><td>1.000</td></tr>
        <tr><td>F1 score</td><td>1.000</td></tr>
        <tr><td>Annotation edit distance</td><td>0.000</td></tr>
      </table>

      <h3>Exon structure comparison</h3>
      <table class="table_normal table_extra_indent">
        <tr><td>reference exons</td><td>38</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>38 (100.0%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>0 (0.0%)</td></tr>
        <tr><td>prediction exons</td><td>38</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>38 (100.0%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>0 (0.0%)</td></tr>
        <tr><td>Sensitivity</td><td>1.000</td></tr>
        <tr><td>Specificity</td><td>1.000</td></tr>
        <tr><td>F1 score</td><td>1.000</td></tr>
        <tr><td>Annotation edit distance</td><td>0.000</td></tr>
      </table>

      <h3>UTR structure comparison</h3>
      <table class="table_normal table_extra_indent">
        <tr><td>reference UTR segments</td><td>12</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>12 (100.0%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>0 (0.0%)</td></tr>
        <tr><td>prediction UTR segments</td><td>12</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>12 (100.0%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>0 (0.0%)</td></tr>
        <tr><td>Sensitivity</td><td>1.000</td></tr>
        <tr><td>Specificity</td><td>1.000</td></tr>
        <tr><td>F1 score</td><td>1.000</td></tr>
        <tr><td>Annotation edit distance</td><td>0.000</td></tr>
      </table>

      <h3>Nucleotide-level comparison</h3>
      <table class="table_wide table_extra_indent">
        <tr><th>&nbsp;</th><th>CDS</th><th>UTRs</th><th>Overall</th></tr>
        <tr><th class="left-align">matching coefficient</th><td>1.000</td><td>1.000</td><td>1.000</td></tr>
        <tr><th class="left-align">correlation coefficient</th><td>1.000</td><td>1.000</td><td>--</td></tr>
        <tr><th class="left-align">sensitivity</th><td>1.000</td><td>1.000</td><td>--</td></tr>
        <tr><th class="left-align">specificity</th><td>1.000</td><td>1.000</td><td>--</td></tr>
        <tr><th class="left-align">F1 score</th><td>1.000</td><td>1.000</td><td>--</td></tr>
        <tr><th class="left-align">annotation edit distance</th><td>0.000</td><td>0.000</td><td>--</td></tr>
      </table>

      <p class="footer">
        Generated by <a href="https://github.com/standage/AEGeAn/releases/tag/v0.16.0">AEGeAn v0.16.0 (stable )</a>.<br />
        Copyright Â© 2016 <a href="http://aegean.readthedocs.io/en/latest/contrib.html">AEGeAn authors</a>.<br />
        See <a href="LICENSE">LICENSE</a> for details.      </p>
</div> </body> </html>
