<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN""http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>SynDivA Report</title><link href="http://twitter.github.com/bootstrap/assets/css/bootstrap.css" rel="stylesheet" /><style type="text/css">body {padding-top: 40px;}.subhead {padding: 40px 0;}.subhead h1 {font-size: 60px;}.fasta {   font-family: Monaco, Menlo, Consolas, "Courier New", monospace;   font-size: 12px;}code.grey{color: #636D71;}</style></head><body><a id="top"></a><div class="navbar navbar-fixed-top"><div class="navbar-inner"><div class="container"><a class="brand" href="#top">SynDivA Report</a><div class="nav-collapse collapse"><ul class="nav"><li><a href="#input">Input data</a></li><li><a href="#analysis">Sequences analysis</a></li><li><a href="#variable">Variable regions analysis</a></li><li><a href="#cluster">Clustering</a></li><li><a href="#stat">Statistics</a></li><li><a href="#annex">Annex</a></li></ul></div></div></div></div><div class="container-fluid"><header class="subhead"><h1>SynDivA Report</h1></header><div class="page-header"><a id="input"></a><h2>Input data</h2></div><p>Input file:<br/><code class="grey">syndiva_datatest.fasta</code></p><p>Number of sequences in input file:<br/><code class="grey">36</code></p><p>Pattern of the sequence bank:<br/><code class="grey">AAAGSSVSSVPTKLEVVAATPTSLLISWDA:4-6:V:1:YYRITYGETGGNSPVQEFTVPG:3:TATISGLSPGVDYTITVYA:11-12:PISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAETVESCLAKSHTENSFTNVWKDDKTLDRYANYE</code></p><p>5' restriction site:<br/><code class="grey">GCGGCCGC</code></p><p>3' restriction site:<br/><code class="grey">GGTACC</code></p><div class="page-header"><a id="analysis"></a><h2>Sequences analysis</h2></div><p>Caption:</p><ul><li class="text-success">Valid sequences that will be part of the next analysis </li><li class="text-warning">Good sequences but will not be part of the next analysis</li><li class="text-error">Rejected sequences</li></ul><table class="table table-striped table-bordered"><tr><th class="text-error">Absence of restriction sites</th><th class="text-error">Incorrect number of nucleotides between the restriction sites</th><th class="text-error">Stop codon <u>inside</u> the area of interest</th><th class="text-warning">Mutation in the conserved regions</th><th class="text-success">Valid sequences</th><th>Amber codon in the sequence (<u>inside</u> the area of interest)</th></tr><tr><td class="text-error">1 sequence(s) (2.78%)</td><td class="text-error">6 sequence(s) (16.67%)</td><td class="text-error">25 sequence(s) (69.44%)</td><td class="text-warning">0 sequence(s) (0.00%)</td><td class="text-success">4 sequence(s) (11.11%)</td><td>0 sequence(s)</td></tr><tr><td class="text-error">XL2-3_PSEXSEQ-REV_34</td><td class="text-error">XL1_22_PSEXSEQ-REV_24<br/>XL1_24_PSEXSEQ-REV_26<br/>XL1_30_PSEXSEQ-REV_30<br/>XL1_4_PSEXSEQ-REV_8<br/>XL2-2_PSEXSEQ-REV_33<br/>XL3-4_PSEXSEQ-REV_40</td><td class="text-error">XL1_10_PSEXSEQ-REV_13<br/>XL1_11_PSEXSEQ-REV_14<br/>XL1_13_PSEXSEQ-REV_16<br/>XL1_14_PSEXSEQ-REV_17<br/>XL1_15_PSEXSEQ-REV_18<br/>XL1_16_PSEXSEQ-REV_19<br/>XL1_17_PSEXSEQ-REV_20<br/>XL1_1_PSEXSEQ-REV_5<br/>XL1_23_PSEXSEQ-REV_25<br/>XL1_25_PSEXSEQ-REV_27<br/>XL1_26_PSEXSEQ-REV_28<br/>XL1_29_PSEXSEQ-REV_29<br/>XL1_33_PSEXSEQ-REV_23<br/>XL1_3_PSEXSEQ-REV_7<br/>XL1_5_PSEXSEQ-REV_9<br/>XL1_6_PSEXSEQ-REV_10<br/>XL1_8_PSEXSEQ-REV_11<br/>XL1_9_PSEXSEQ-REV_12<br/>XL2-1_PSEXSEQ-REV_32<br/>XL2-4_PSEXSEQ-REV_35<br/>XL2-5_PSEXSEQ-REV_36<br/>XL3-1_PSEXSEQ-REV_37<br/>XL3-2_PSEXSEQ-REV_38<br/>XL3-3_PSEXSEQ-REV_39<br/>XL3-5_PSEXSEQ-REV_41</td><td class="text-warning"></td><td class="text-success">XL1_12_PSEXSEQ-REV_15<br/>XL1_18_PSEXSEQ-REV_21<br/>XL1_19_PSEXSEQ-REV_22<br/>XL1_2_PSEXSEQ-REV_6</td><td></td></tr></table><div class="page-header"><a id="variable"></a><h2>Variable regions analysis</h2></div><p>The following group of sequences are identical clones on the variable regions:</p><p>No clone was found.</p><p>Here's the distribution of the repeated sequences in variable regions:</p><table class="table table-striped table-bordered"><thead><tr><th>Variable region</th><th>Repeated sequence</th><th>Number of occurrences (percentage of valid sequences)</th></tr></thead><tbody><tr><td rowspan="1">3</td><td>YSY</td><td>2 (50.00%)</td></tr></tbody></table><div class="page-header"><a id="cluster"></a><h2>Clustering</h2></div><p>The following clusters were generated by MCL:</p><div class="row-fluid"><div class="span6"><pre>4 sequences (100.00% of valid sequences)<br/>XL1_12_PSEXSEQ-REV_15<br/>XL1_18_PSEXSEQ-REV_21<br/>XL1_19_PSEXSEQ-REV_22<br/>XL1_2_PSEXSEQ-REV_6
</pre></div></div><div class="page-header"><a id="stat"></a><h2>Statistics</h2></div><p>Here's some statistics about the valid sequences:</p><p>Mean for the pairwise alignement scores: 20.01<br/>Standard deviation: 2.93</p><div class="row-fluid"><div class="span6"><img src="distri.png" alt="Distribution of the pairwise alignment score"></div><div class="span6"><table class="table table-striped table-bordered"><thead><tr><th>Pairwise Alignment Score</th><th>Number of occurrences</th></tr></thead><tbody><tr><td>16.67</td><td>1</td></tr><tr><td>17.24</td><td>1</td></tr><tr><td>19.23</td><td>1</td></tr><tr><td>19.35</td><td>1</td></tr><tr><td>22.58</td><td>1</td></tr><tr><td>25.00</td><td>1</td></tr></tbody></table></div></div><div class="page-header"><a id="annex"></a><h2>Annex</h2></div><p><strong>Valid protein sequences</strong> in FASTA format:</p><textarea class="span8 fasta" type="text" rows="20" readonly="readonly">>XL1_12_PSEXSEQ-REV_15
AAAGSSVSSVPTKLEVVAATPTSLLISWDANLPNSAVHYYRITYGETGGNSPVQEFTVPGSSYTATISGLSPGVDYTITV
YAHNDCNTIVATCPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAR
>XL1_18_PSEXSEQ-REV_21
AAAGSSVSSVPTKLEVVAATPTSLLISWDAIYRPVAYYRITYGETGGNSPVQEFTVPGYSYTATISGLSPGVDYTITVYA
PIGHFDVLVTIPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRD
>XL1_19_PSEXSEQ-REV_22
AAAGSSVSSVPTKLEVVAATPTSLLISWDASIFRGIVLYYRITYGETGGNSPVQEFTVPGYSYTATISGLSPGVDYTITV
YAHGTSFSAISCAFPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAM
>XL1_2_PSEXSEQ-REV_6
AAAGSSVSSVPTKLEVVAATPTSLLISWDAPLGDVFYYRITYGETGGNSPVQEFTVPGSYYTATISGLSPGVDYTITVYA
TGARTSADGFIPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAR
</textarea><p>Multiple sequence alignment of the <strong>valid sequences</strong> generated by Clustal Omega:</p><textarea class="span8 fasta" type="text" rows="20" readonly="readonly">CLUSTAL O(1.2.4) multiple sequence alignment


XL1_12_PSEXSEQ-REV_15      AAAGSSVSSVPTKLEVVAATPTSLLISWDANLPNSAVHYYRITYGETGGNSPVQEFTVPG
XL1_18_PSEXSEQ-REV_21      AAAGSSVSSVPTKLEVVAATPTSLLISWDAIY--RPVAYYRITYGETGGNSPVQEFTVPG
XL1_19_PSEXSEQ-REV_22      AAAGSSVSSVPTKLEVVAATPTSLLISWDASIFRGIVLYYRITYGETGGNSPVQEFTVPG
XL1_2_PSEXSEQ-REV_6        AAAGSSVSSVPTKLEVVAATPTSLLISWDAPL--GDVFYYRITYGETGGNSPVQEFTVPG
                           ******************************      * **********************

XL1_12_PSEXSEQ-REV_15      SSYTATISGLSPGVDYTITVYAHNDCNTI-VATCPISINYRTGTGGSGGSHHHHHHHHGG
XL1_18_PSEXSEQ-REV_21      YSYTATISGLSPGVDYTITVYAPIGHFDV-LVTIPISINYRTGTGGSGGSHHHHHHHHGG
XL1_19_PSEXSEQ-REV_22      YSYTATISGLSPGVDYTITVYAHGTSFSAISCAFPISINYRTGTGGSGGSHHHHHHHHGG
XL1_2_PSEXSEQ-REV_6        SYYTATISGLSPGVDYTITVYATGARTSA-DGFIPISINYRTGTGGSGGSHHHHHHHHGG
                             ********************            **************************

XL1_12_PSEXSEQ-REV_15      SYPYDVPDYAPRVG*GRVGSKDIRAR
XL1_18_PSEXSEQ-REV_21      SYPYDVPDYAPRVG*GRVGSKDIRD-
XL1_19_PSEXSEQ-REV_22      SYPYDVPDYAPRVG*GRVGSKDIRAM
XL1_2_PSEXSEQ-REV_6        SYPYDVPDYAPRVG*GRVGSKDIRAR
                           ************************  
</textarea><p><strong>Protein sequences with an incorrect number of nucleotides between the restriction sites</strong> in FASTA format:</p><textarea class="span8 fasta" type="text" rows="20" readonly="readonly">>XL1_22_PSEXSEQ-REV_24
AAAGSSVSSVPTKLEVVAATPTSLLISWDALITFVAYYRITYGETGGTPRFRNSPSRVLILPQRFPACHRVSTIPLRFTL
PTPPTTTTSSSRFLSITVPVPAVRAVAITITITIMVVRIRMTCQTMHHAWARDAWDPKISEPI
>XL1_24_PSEXSEQ-REV_26
AAAGSSVSSVPTKLEVVAATPTSLLISWDAHSSPDYVCYYRITYGETGVTPGSGIHRPG*FLNRNDFRPVTGCRLYHYGL
RFELLTYLLIEL*SDFYQLPYRYRRFGR*PSPSPSPSWWFVSV*RARLCTTRGLGTRGIQRYQSK
>XL1_30_PSEXSEQ-REV_30
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPSAATPTTRFL
SITVPVPAVRAVAITITITIMVVRIRMTCQTMHHAWARDAWDPKIS
>XL1_4_PSEXSEQ-REV_8
AAAGSSVSSVPTKLEVVAATPTSLLISWDAPFTDHCVVYYRITYGETGGNPRFRNSPSRVLIIPQRFPACHRVSTIPLRF
TLTTPAATTTAPTPRFLSITVPVPAVRAVAITITITIMVVRIRMTCQTMHHAWARDAWDPKISEQS
>XL2-2_PSEXSEQ-REV_33
AAAGSSVSSVPTKLEVVAATPTSLLISWDACHTFVNYYRITYGETGGNSPVQEFTRPGLFLYRNDFRPVTGCRLYHYGLR
*PLPPLLRLPRPDFYQLPYRYRRFGR*PSPSPSPSWWFVSV*RARLCTTRGLGTRGIQRYPE
>XL3-4_PSEXSEQ-REV_40
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTSRVLIIPQRFPACHRVSTIPLRFTL
PASTTASSTAADFYQLPYRYRRFGR*PSPSPSPSWWFVSV*RARLCTTRGLGTRGIQRYQSER
</textarea><p><strong>Protein sequences with a stop codon</strong> in FASTA format:</p><textarea class="span8 fasta" type="text" rows="20" readonly="readonly">>XL1_10_PSEXSEQ-REV_13
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAT
>XL1_11_PSEXSEQ-REV_14
AAAGSSVSSVPTKLEVVAATPTSLLISWDAHGNCVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAE
>XL1_13_PSEXSEQ-REV_16
AAAGSSVSSVPTKLEVVAATPTSLLISWDAPRSFVRYYRITYGETGGNSPVQEFTVPGSYSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAK
>XL1_14_PSEXSEQ-REV_17
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSYTATISGLSPGVDYTITVYA
RYFIYSYISHSTPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAK
>XL1_15_PSEXSEQ-REV_18
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRC
>XL1_16_PSEXSEQ-REV_19
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGYYSTATISGLSPGVDYTITVYA
CGGVNANSSDCFPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAK
>XL1_17_PSEXSEQ-REV_20
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SVPIHFSCRGCPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIS
>XL1_1_PSEXSEQ-REV_5
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIP
>XL1_23_PSEXSEQ-REV_25
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSYYTATISGLSPGVDYTITVYA
GDISDNPFSRCPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAN
>XL1_25_PSEXSEQ-REV_27
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAR
>XL1_26_PSEXSEQ-REV_28
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKRYP
>XL1_29_PSEXSEQ-REV_29
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAP
>XL1_33_PSEXSEQ-REV_23
AAAGSSVSSVPTKLEVVAATPTSLLISWDASQSNVSYYRITYGETGGNSPVQEFTVPGCYSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAEN
>XL1_3_PSEXSEQ-REV_7
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAS
>XL1_5_PSEXSEQ-REV_9
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAN
>XL1_6_PSEXSEQ-REV_10
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGYYSTATISGLSPGVDYTITVYA
NATVSNFHPINSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDFRAK
>XL1_8_PSEXSEQ-REV_11
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAK
>XL1_9_PSEXSEQ-REV_12
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVCYYRITYGETGGNSPVQEFTVPGSYSTATISGLSPGVDYTITVYA
SLVNDDVHAVHHPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIQS
>XL2-1_PSEXSEQ-REV_32
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKEIRAM
>XL2-4_PSEXSEQ-REV_35
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAN
>XL2-5_PSEXSEQ-REV_36
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSYYTATISGLSPGVDYTITVYA
SGYCVIVANSYPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIP
>XL3-1_PSEXSEQ-REV_37
AAAGSSVSSVPTKLEVVAATPTSLLISCDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRSR
>XL3-2_PSEXSEQ-REV_38
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAR
>XL3-3_PSEXSEQ-REV_39
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAS
>XL3-5_PSEXSEQ-REV_41
AAAGSSVSSVPTKLEVVAATPTSLLISWDAS*SSVSYYRITYGETGGNSPVQEFTVPGSSSTATISGLSPGVDYTITVYA
SSS*RISSSSSSPISINYRTGTGGSGGSHHHHHHHHGGSYPYDVPDYAPRVG*GRVGSKDIRAT
</textarea></div></body></html>