<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html xmlns:cis="http://zlab.bu.edu/schema/cisml" xmlns:fimo="http://noble.gs.washington.edu/schema/cisml" xmlns:mem="http://noble.gs.washington.edu/meme">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="UTF-8">
<title>FIMO Results</title>
<style type="text/css">
td.left {text-align: left;}
td.right {text-align: right; padding-right: 1cm;}
</style>
</head>
<body color="#000000">
<a name="top_buttons"></a>
<hr>
<table summary="buttons" align="left" cellspacing="0">
<tr>
<td bgcolor="#00FFFF"><a href="#database_and_motifs"><b>Database and Motifs</b></a></td>
<td bgcolor="#DDFFDD"><a href="#sec_i"><b>High-scoring Motif Occurences</b></a></td>
<td bgcolor="#DDDDFF"><a href="#debugging_information"><b>Debugging Information</b></a></td>
<td bgcolor="#FFDD00"><a href="fimo.tsv""><b>Results in TSV Format</b></a></td>
<td bgcolor="#FFDDFF"><a href="fimo.gff""><b>Results in GFF3 Format</b></a></td>
</tr>
</table>
<center><big><b>FIMO - Motif search tool</b></big></center>
FIMO version 5.0.5, (Release date: Mon Mar 18 20:12:19 2019 -0700)
For further information on how to interpret these results please access <a href="http://meme-suite.org/doc/fimo-output-format.html">http://meme-suite.org/doc/fimo-output-format.html</a>.<br>
To get a copy of the FIMO software please access <a href="http://meme-suite.org">http://meme-suite.org</a>
<p>If you use FIMO in your research, please cite the following paper:<br>
Charles E. Grant, Timothy L. Bailey, and William Stafford Noble,
"FIMO: Scanning for occurrences of a given motif",
<i>Bioinformatics</i>, <b>27</b>(7):1017-1018, 2011.
<center><big><b><a name="database_and_motifs">DATABASE AND MOTIFS</a></b></big></center>
  Database contains 1 sequences, 16569 residues
  <table>
    <thead>
    </thead>
    <tbody>
    </tbody>
  </table>
A 0.278 C 0.222 G 0.222 T 0.278 </p>
<center><big><b><a name="sec_i">SECTION I: HIGH-SCORING MOTIF OCCURENCES</a></b></big></center>
There were 11 motif occurences with a p-value less than 0.0001.
The full set of motif occurences can be seen in the
TSV (tab-delimited values) output file
the GFF3 file 
which may be suitable for uploading to the 
(assuming the FASTA input sequences included genomic coordinates in UCSC or Galaxy format),
or the XML file 
The <i>p</i>-value of a motif occurrence is defined as the
probability of a random sequence of the same length as the motif
matching that position of the sequence with as good or better a score.
The score for the match of a position in a sequence to a motif
is computed by summing the appropriate entries from each column of
the position-dependent scoring matrix that represents the motif.
The q-value of a motif occurrence is defined as the
false discovery rate if the occurrence is accepted as significant.
<li>The table is sorted by increasing <i>p</i>-value.</li>
<th>Motif ID</th>
<th>Alt ID</th>
<th>Sequence Name</th>
<th>Strand</th>
<th>Start</th>
<th>End</th>
<th>p-value</th>
<th>q-value</th>
<th>Matched Sequence</th>
    <tr>
      <td style="text-align:left;">CCAGCAY</td>
      <td style="text-align:left;">DREME-5</td>
      <td style="text-align:left;">chrM</td>
      <td style="text-align:center;">+</td>
      <td style="text-align:left;">510</td>
      <td style="text-align:left;">516</td>
      <td style="text-align:left;">4.15e-05</td>
      <td style="text-align:left;">0.683</td>
      <td style="text-align:left;font-size:x-large;font-family:monospace;">CCAGCAC</td>
      <td style="text-align:left;">2093</td>
      <td style="text-align:left;">2099</td>
      <td style="text-align:left;">8.2e-05</td>
      <td style="text-align:left;">0.332</td>
      <td style="text-align:left;font-size:x-large;font-family:monospace;">ACTAACA</td>
   </tr>
<hr>
<center><big><b><a name="debugging_information">DEBUGGING INFORMATION</a></b></big></center>
Command line:
Settings:
This information can be useful in the event you wish to report a
problem with the FIMO software.
</body>
</html>
